BlockChainRunner.java

```diff
@@ -0,0 +1,354 @@
+            new Miner("Miner 6", transactionPool, 5, blockQueue, threadSyncSemaphore),
+            new Miner("Miner 7", transactionPool, 6, blockQueue, threadSyncSemaphore),
+            new Miner("Miner 8", transactionPool, 7, blockQueue, threadSyncSemaphore),
+            new Miner("Miner 9", transactionPool, 8, blockQueue, threadSyncSemaphore),
+            new Miner("Miner 10", transactionPool, 9, blockQueue, threadSyncSemaphore),
+            new Miner("Miner 11", transactionPool, 10, blockQueue, threadSyncSemaphore)
+        };
+        startMinerThreads(miners);
+
+
+        while (true) {
+            generatePendingTransactions();
+            addPendingTransactionsToPool();
+
+            if (transactionPool.isEmpty()) break;
+
+            threadSyncSemaphore.acquire(miners.length);
+            blockQueue.clear();
+            startMiningNextBlock(blockChain.get(blockChain.size() - 1).getHash(), REGULAR_BLOCK_DIFFICULTY, miners);
+
+            Pair<Block, Integer> minedValue = blockQueue.take();
+            threadSyncSemaphore.acquire(miners.length);
+            pauseAllMiners(miners);
+            log(miners[minedValue.getRight()] + " mined block " + blockChain.size() + ": " + minedValue.getLeft());
+            addBlockToBlockChain(minedValue.getLeft());
+        }
+
+        log("Block chain simulation ended");
+        terminateAllMiners(miners);
+    }
+
+    private void startMinerThreads(Miner... miners) {
+        for (Miner miner : miners) {
+            new Thread(miner, miner.getName() + " thread").start();
+        }
+    }
+
+    private void startMiningNextBlock(String prevBlockHash, int difficulty, Miner... miners) {
+        for (Miner miner : miners) {
+            miner.mineNextBlock(prevBlockHash, difficulty);
+        }
+    }
+
+    private void pauseAllMiners(Miner... miners) {
+        for (Miner miner : miners) {
+            miner.pauseMining();
+        }
+    }
+
+    private void terminateAllMiners(Miner... miners) {
+        for (Miner miner : miners) {
+            miner.stopExecution();
+        }
+    }
+
+    private void mineGenesisBlock() throws InterruptedException {
+        List<TransactionOutput> transactionOutputs = new ArrayList<>();
+
+        users.forEach(user -> {
+            transactionOutputs.add(
+                TransactionOutput.builder().destPublicKey(user.getPublicKey())
+                    .value(ThreadLocalRandom.current().nextInt(1, 1000000))
+                    .build()
+            );
+        });
+
+        transactionPool.add(new Transaction(new ArrayList<>(), transactionOutputs));
+
+        BlockingQueue<Pair<Block, Integer>> genesisQueue = new LinkedBlockingQueue<>();
+        Miner genesisMiner = new Miner("Genesis miner", transactionPool, 0, genesisQueue, new Semaphore(10));
+        startMinerThreads(genesisMiner);
+        startMiningNextBlock(BLOCK_ZERO_HASH, GENESIS_BLOCK_DIFFICULTY, genesisMiner);
+        Block genesisBlock = genesisQueue.take().getKey();
```

Comment:
Seems to be a violation of Law of Demeter, since genesisQueue.take().getKey() is breaking the initial principle.

Would be suitable to reform the code, where:

`
Block genesisBlock = genesisQueue.take()
addBlockToBlockChain(genesisBlock.getKey());
log("mined genesis block " + genesisBlock.getKey());

`

Reply:
Addressed